Add Withings support

vangorra opened the pull request:
## Description:
Adding support for connecting to Withings apis to integrate health data into home assistant.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8816

## Example entry for `configuration.yaml` (if applicable):
```yaml
# Example configuration.yaml entry
withings:
    client_id: <client id from previous step>
    client_secret: <consumer secret from previous step>
    profiles: 
        - <name of user profile>
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) [PR](https://github.com/home-assistant/home-assistant.io/pull/8816)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

vangorra committed:
Initial commit

vangorra committed:
Merge branch 'dev' into vangorra_withings

vangorra committed:
Adding a few tests.

vangorra committed:
Adding view test.

vangorra committed:
Adding data manager tests.
Putting init tests into a class.

vangorra committed:
Removing unused properties.

vangorra committed:
More tests.

vangorra committed:
Updating nokia module.

vangorra committed:
Finishing tests.
Adding sleep summary data.
Adding current sleep state.

vangorra committed:
Removing pint requirement

vangorra committed:
Addressing all pylint and flake8 warnings.

vangorra committed:
Addressing all pylint and flake8 warnings.

vangorra committed:
Adding dependencies

vangorra committed:
Updating title.

vangorra committed:
Trying out url parameters for slug.

vangorra committed:
Redirect to root of ha

vangorra committed:
Disabling auth for callback views.

vangorra committed:
Adding sleep icon.

vangorra committed:
Renaming sleep summary constant values.
Setting unit of measurements for sleep state to none

vangorra committed:
Cleanups.

vangorra committed:
Adding sensor to coverage config.

homeassistant commented:
Hi @vangorra, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#21625).

Once you do that we will be able to review and accept this pull request.

Thanks!

vangorra committed:
Addressing automated CR feedback.

vangorra committed:
Fixing flake 8 warnings.

vangorra committed:
Fixing test errors

vangorra committed:
Moving files around to match new directory structure.

vangorra committed:
Regenerating the requirements files.

vangorra committed:
Adding entity picture to configuring view.

vangorra committed:
Cleaning up after tests.

balloob commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,1057 @@
+    await async_initialize(configuring, creds)
+
+    _LOGGER.debug('Finishing request.')
+    configuring.hass.components.configurator.async_request_done(
```
We should no longer use the configurator. We should use a config flow instead.

vangorra replied:
Are there any good examples available for using a ConfigFlow? I based this platform on other components I found in source but didn't see anything about config flow.

cgtobi replied:
A good start would be the architecture docs. https://developers.home-assistant.io/docs/en/config_entries_config_flow_handler.html

MartinHjelmare replied:
Look at nest and point components.

vangorra committed:
Code refactor to use new directory structure and config flows.

vangorra committed:
More restructuring.
Debugging.

vangorra committed:
Adding config unload.

vangorra committed:
Merge branch 'dev' into vangorra_withings

# Conflicts:
#	requirements_all.txt
#	requirements_test.txt
#	requirements_test_all.txt

vangorra committed:
Adding workaround for unit conversion.
Fixing tests.
Updating requirements.

vangorra committed:
Fixing flake8 styles.
Adding more tests.
Giving Entity the ability to ignore temperature conversions.

vangorra committed:
Trying to get flake on circleci to pass.

vangorra committed:
Adding entry to code owners.

vangorra committed:
Merge branch 'dev' into vangorra_withings

vangorra committed:
Updating requirements.

vangorra committed:
Adding stone for all weight measurements.

vangorra committed:
Fixing static analysis.

vangorra committed:
Fixing tests so they worth with python 3.5.5

vangorra committed:
Adding withings to coverage config.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+                ),
+                vol.Optional(
+                    const.MEASURES,
+                    default=list(WITHINGS_MEASUREMENTS_MAP.keys())
```
```suggestion
                    default=list(WITHINGS_MEASUREMENTS_MAP)
```

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+                    vol.Length(min=1),
+                    [vol.All(
+                        cv.string,
+                        vol.In(list(WITHINGS_MEASUREMENTS_MAP.keys()))
```
```suggestion
                        vol.In(list(WITHINGS_MEASUREMENTS_MAP))
```

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType):
+    """Set up the Withings component."""
+    conf = config.get(DOMAIN, None)
```
```suggestion
    conf = config.get(DOMAIN)
```

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+        _LOGGER.debug("No withing config was provided.")
+        return True
+
+    _LOGGER.debug("Saving component config for later.")
```
Remove this.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+        const.CONFIG: conf
+    }
+
+    _LOGGER.debug("Determining the base url to use for callbacks.")
```
Remove this.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+        hass.config.api.base_url
+    ).rstrip('/')
+
+    _LOGGER.debug("Setting up configuration flow data.")
```
Remove this.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+            profile
+        )
+
+    _LOGGER.debug("Initializing configuration flow.")
```
Remove this.

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.helpers.typing import HomeAssistantType, ConfigType
+from homeassistant.helpers import config_validation as cv
+from homeassistant.components.withings import (
```
All internal integration package imports should be relative to allow custom component override.
```py
from . import ...
```

MartinHjelmare commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,136 @@
+https://home-assistant.io/components/withings/
+"""
+import voluptuous as vol
+from homeassistant.config_entries import ConfigEntry
```
Please group and sort imports according to [PEP8](https://www.python.org/dev/peps/pep-0008/#imports).

Ie, add a blank line here.

MartinHjelmare commented on .coveragerc:
```diff
@@ -654,6 +654,7 @@ omit =
     homeassistant/components/whois/sensor.py
     homeassistant/components/wink/*
     homeassistant/components/wirelesstag/*
+    homeassistant/components/withings/*
```
Don't exclude modules from coverage if their coverage is good. The `config_flow.py` module must not be excluded and must have 100% coverage.

vangorra replied:
Opps. I overlooked this, I thought it was inclusions.

MartinHjelmare reviewed (changes requested):
Please be more conservative with debug logging. Don't log internal logic, unless there's some critical calculation done and we should be able to check the result. Logging that we're doing something without providing any state data is rarely useful. It's ok to log an api response.

I haven't looked at the config flow yet.

vangorra commented:
@MartinHjelmare Clarifying question about logging. In my experience, people don't enable debug/trace logs unless there is a problem. Further one needs debug/trace logging to field tickets for un-reproducible events. So verbosity is key. My professional rule of thumb tends to be that most comments can be replaced with debug/trace log statements. Although verbose, verbose debug logging will provide value when fielding tickets related to this component. Granted a trace log level would be more appropriate, is there one available? I didn't see it.

MartinHjelmare commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,645 @@
+        async_add_entities: types.Callable[[types.List[Entity], bool], None]
+):
+    """Set up the sensor config entry."""
+    import nokia
```
Scratch that. It's possible to move the imports to the top but currently integrations with config flow will install requirements already at config flow registration, so will mean possibly unnecessary libraries installed.

We're fixing this.

MartinHjelmare commented:
I don't think there's a trace level. Here are the standard levels:
https://docs.python.org/3/library/logging.html#logging-levels

vangorra committed:
Fixing imports and removing entity class modifications.

vangorra committed:
Refactoring tests to use pytest fixtures.

vangorra committed:
More tests and service checks.

vangorra committed:
Addressing remaining PR feedback.

vangorra committed:
Bug fixing.

vangorra commented on homeassistant/helpers/entity.py:
```diff
@@ -189,6 +189,16 @@ def context_recent_time(self):
         return timedelta(seconds=5)
 
+    @property
+    def convert_units(self) -> bool:
```
Added as #23499

vangorra replied:
https://github.com/home-assistant/architecture/issues/217

vangorra committed:
Fixing test compatibility with python 3.5.5

MartinHjelmare commented on homeassistant/components/withings/common.py:
```diff
@@ -0,0 +1,184 @@
+from . import const
+
+_LOGGER = logging.getLogger(const.LOG_NAMESPACE)
+NOT_AUTHENTICATED_ERROR = re.compile(".*(Error Code (100|101|102|200|401)|Missing access token parameter).*", re.IGNORECASE)  # pylint: disable=line-too-long  # noqa: E501
```
Please break the line instead of disabling too long line. Break the line after the first parenthesis.

MartinHjelmare commented on homeassistant/components/withings/common.py:
```diff
@@ -0,0 +1,184 @@
+    @Throttle(const.SCAN_INTERVAL)
+    async def async_check_authenticated(self):
+        """Check if the user is authenticated."""
+        async def function():
```
Why are these functions coroutines? They don't seem to await anything. If we don't need them to be coroutines, it's better to let them be regular callbacks.

MartinHjelmare commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,635 @@
+
+    measures = hass.data.get(const.DOMAIN, {}) \
+        .get(const.CONFIG, {}) \
+        .get(const.MEASURES, list(WITHINGS_MEASUREMENTS_MAP.keys()))
```
Remove `.keys()`.

MartinHjelmare commented on tests/components/withings/test_sensor.py:
```diff
@@ -0,0 +1,767 @@
+):
+    """Test method."""
+    attribute = WITHINGS_MEASUREMENTS_MAP[const.MEAS_WEIGHT_KG]
+    sensor = WithingsHealthSensor(
```
Don't assert things about the entity directly. Set up the component which should set up the platform. Use `async_setup_component`. Let the platform add an entity. Then from the core state machine, get the state of the entity and assert the state and the state attributes.

The state is the truth of the entity.

Writing the platform tests like this will make them more robust.

vangorra replied:
These sensors don't get created until after the config flow and config entry hooks are fired. That whole process is covered in other tests. These tests focus primarily on confirming the unit conversion code is working properly.

MartinHjelmare replied:
The point is that we shouldn't interact with the entity directly in tests about the platform. We should use the state as the truth of the entity.

vangorra replied:
I don't understand what is being asked here. The test in question "test_health_sensor_properties" only tests the mapping of properties when the entity is created. It doesn't use any platform fixtures. I'm not sure how this applies when you say "shouldn't interact with the entity directly in tests about the platform".

MartinHjelmare replied:
I'm asking you to do this instead:
> Set up the component which should set up the platform. Use async_setup_component. Let the platform add an entity. Then from the core state machine, get the state of the entity and assert the state and the state attributes.

vangorra replied:
This test's purpose is to confirm the variables passed to the constructor are correctly provided as properties with the same name.

Using async_setup_component sounds like a lot of overhead for what I'm trying to test here.

Using async_setup_component means I have to supply a component config for withings and it's dependencies. From there withings' async_setup is called which I have to mock configs so the setup is forwarded to the sensor platform. Once there, over a dozen WithingsSensor instances are created.

I already have tests that cover these platform interactions.

MartinHjelmare replied:
Then we can remove this test. We shouldn't write tests that enforce rigidity in the entity implementation. We should be able to refactor the platform without changing tests, as long as we don't make breaking changes. Breaking changes will occur if we change what state the entity reports.

MartinHjelmare replied:
Make sure we test all the state and state attributes of the entity in some tests, as I've described above.

vangorra replied:
That makes sense. Looking into it, I already have coverage for this. I'll just remove this unnecessary test.

vangorra committed:
Addressing code review feedback.

vangorra committed:
Merge branch 'dev' into vangorra_withings

vangorra committed:
Setting parallel updates constant.

MartinHjelmare commented on homeassistant/components/withings/sensor.py:
```diff
@@ -14,6 +14,11 @@
+# There's only 3 calls (per profile) made to the withings api every 5
+# minutes (see throttle values). This component wouldn't benefit
+# much from parallel updates.
+PARALLEL_UPDATES = 0
```
Set it to 1 if we only want 1 update at a time.

Setting it to 0 will mean limitless parallel updates.

vangorra committed:
Fixing parallel updates value.

vangorra committed:
Removing offending test. Other tests already cover the functionality.

MartinHjelmare commented on tests/components/withings/test_sensor.py:
```diff
@@ -0,0 +1,748 @@
+    )
+
+    await sensor.async_update()
+    assert sensor.state == expected, \
```
This test helper should be rewritten as we've discussed. It should set up the component which sets up the platform, adds the entity, gets the state from the core state machine and then asserts the state and state attributes of the entity.

vangorra replied:
Sounds like higher level integration tests are what's desired here. I've updated the tests accordingly.

vangorra committed:
Using integration style tests for entities.
Some bug fixing.

vangorra committed:
Merge branch 'dev' into vangorra_withings

vangorra committed:
Fixing bugs around throttling.
Adding icon to auto temperature.

MartinHjelmare commented on homeassistant/components/withings/config_flow.py:
```diff
@@ -192,25 +197,45 @@ class WithingsAuthCallbackView(HomeAssistantView):
-                    const.DOMAIN,
-                    context={'source': const.CODE},
-                    data={
+                hass.config_entries.flow.async_configure(
```
Why did we change to `async_configure`?

vangorra replied:
So it would reuse the existing flow_id rather than starting a new flow. I thought that may work better in the future. Although there isn't any docs on the subject so it's hard to say.

MartinHjelmare replied:
How does it work now?

vangorra replied:
It works the same as before, the integration gets setup normally. Since the click link step has no way for the user to advance other than clicking the link, I was hoping that by doing it this way, the UI would advance the click link step. It did not.

MartinHjelmare commented on homeassistant/components/withings/sensor.py:
```diff
@@ -479,7 +460,11 @@ def convert_units(self):
+                "Provided data is None. Setting state to %s",
+                const.STATE_UNKNOWN
+            )
+            self._state = const.STATE_UNKNOWN
```
Use `None` for unknown state.

MartinHjelmare commented on tests/components/withings/test_config_flow.py:
```diff
@@ -162,10 +161,9 @@ def test_flow_handler_get_auth_client(
-        'profile': 'Person 1',
-    }
+    assert result['description_placeholders']
+    assert result['description_placeholders']['authorization_url'].startswith('https://account.withings.com/oauth2_user/authorize2?response_type=code&client_id=my_client_id&redirect_uri=http%3A%2F%2Flocalhost%2Fapi%2Fwithings%2Fcallback%2Fperson_1&scope=user.info%2Cuser.metrics%2Cuser.activity&state=')  # pylint: disable=line-too-long  # noqa: E501
```
Please break long lines.

MartinHjelmare commented on tests/components/withings/test_sensor.py:
```diff
@@ -1,748 +1,264 @@
-        )
+def get_entity_id(measure, profile):
+    """Get an entity id for a measure and profile."""
+    return 'sensor.%s_%s_%s' % (
```
Use new style string formatting, with `.format`,  instead of old style.

MartinHjelmare commented on tests/components/withings/test_sensor.py:
```diff
@@ -1,748 +1,264 @@
 
-    return api
 
+async def assert_state_equals(
```
This doesn't need to be a coroutine. There's no `await` in the function. Please make it a regular function.

vangorra committed:
Addressing code review feedback.

MartinHjelmare commented on tests/components/withings/test_config_flow.py:
```diff
@@ -0,0 +1,396 @@
+    client = flow_handler.get_auth_client('Person 1')
+    assert client.client_id == 'my_client_id'
+    assert client.consumer_secret == 'my_client_secret'
+    assert client.callback_uri == 'http://localhost/api/withings/callback/person_1'  # pylint: disable=line-too-long  # noqa: E501
```
```suggestion
    assert client.callback_uri == (
        'http://localhost/api/withings/callback/person_1')
```

MartinHjelmare commented on tests/components/withings/test_config_flow.py:
```diff
@@ -0,0 +1,396 @@
+            'Person 1'
+        )
+        client = flow_handler.get_auth_client('Person 1')
+        assert client.callback_uri == 'https://vghome.duckdns.org/api/withings/callback/person_1'  # pylint: disable=line-too-long  # noqa: E501
```
```suggestion
        assert client.callback_uri == (
            'https://vghome.duckdns.org/api/withings/callback/person_1')
```

MartinHjelmare commented on tests/components/withings/test_config_flow.py:
```diff
@@ -0,0 +1,396 @@
+    placeholders = result['description_placeholders']
+    assert placeholders
+    assert placeholders['authorization_url'].startswith(
+        'https://account.withings.com/oauth2_user/authorize2?response_type=code&client_id=my_client_id&redirect_uri=http%3A%2F%2Flocalhost%2Fapi%2Fwithings%2Fcallback%2Fperson_1&scope=user.info%2Cuser.metrics%2Cuser.activity&state='  # pylint: disable=line-too-long  # noqa: E501
```
```suggestion
        'https://account.withings.com/oauth2_user/authorize2?'
        'response_type=code&client_id=my_client_id&redirect_uri='
        'http%3A%2F%2Flocalhost%2Fapi%2Fwithings%2Fcallback%2Fperson_1&scope='
        'user.info%2Cuser.metrics%2Cuser.activity&state='
```

MartinHjelmare commented on tests/components/withings/test_config_flow.py:
```diff
@@ -0,0 +1,396 @@
+        assert result['type'] == data_entry_flow.RESULT_TYPE_FORM
+        assert result['step_id'] == 'auth'
+        assert result['description_placeholders']
+        assert result['description_placeholders']['authorization_url'].startswith('https://account.withings.com/oauth2_user/authorize2?response_type=code&client_id=my_client_id&redirect_uri=http%3A%2F%2Flocalhost%2Fapi%2Fwithings%2Fcallback%2Fperson_1&scope=user.info%2Cuser.metrics%2Cuser.activity&state=')  # pylint: disable=line-too-long  # noqa: E501
```
```suggestion
        assert result['description_placeholders'][
            'authorization_url'].startswith(
                'https://account.withings.com/oauth2_user/authorize2?'
                'response_type=code&client_id=my_client_id&redirect_uri='
                'http%3A%2F%2Flocalhost%2Fapi%2Fwithings%2Fcallback%2F'
                'person_1&scope='
                'user.info%2Cuser.metrics%2Cuser.activity&state=')
```

MartinHjelmare commented on tests/components/withings/conftest.py:
```diff
@@ -0,0 +1,378 @@
+            nokia_api_get_sleep_summary_patch.start()
+
+        data_manager_get_throttle_interval_patch = asynctest.patch(
+            'homeassistant.components.withings.common.WithingsDataManager.get_throttle_interval',  # pylint: disable=line-too-long  # noqa: E501
```
```suggestion
            'homeassistant.components.withings.common.'
            'WithingsDataManager.get_throttle_interval',
```

vangorra committed:
Addressing code review feedback.

vangorra committed:
Better code coverage.
Better tests component setup/teardown.

vangorra committed:
Merge branch 'dev' into vangorra_withings

# Conflicts:
#	homeassistant/config_entries.py

vangorra committed:
Adding new manifest entry and generating config flows.

vangorra commented:
Merged with latest dev, updated manifest to provide config_flow and regenerated the config_flows. My component no longer appears on the integrations page. What am I missing?

MartinHjelmare commented:
Try generating the english translation file, by running the translation script.

https://developers.home-assistant.io/docs/en/internationalization_backend_localization.html#translation-strings

vangorra commented:
Thanks @MartinHjelmare It's working now.

MartinHjelmare commented:
I think this PR mostly looks good. Since it's very big, I would like a second pair of eyes from a member before merging. We usually try to keep PRs way smaller than this. It is a lot of tests that take lines though.

MartinHjelmare commented on tests/components/withings/conftest.py:
```diff
@@ -0,0 +1,379 @@
+
+        async_create_task_patch = asynctest.patch.object(
+            hass,
+            'async_create_task',
```
I'm a bit hesitant to patch `hass.async_create_task`. Can we find a way without patching the core of home assistant?

Eg in the smartthings tests I think we mock the config entry and forward the entry to the platform directly from the component without calling `async_setup_component` directly. Maybe something like that is possible?

https://github.com/home-assistant/home-assistant/blob/94a2fd542ea47abf59d2e3a21f17637a3d9af92f/tests/components/smartthings/conftest.py#L27-L45

vangorra replied:
This patch is only used once (as a wrapper) to retrieve the flow_id of the config flow and then it's released. I experimented with better ways to retrieve the flow_id but had no success. Do you have suggestions?

MartinHjelmare replied:
Do we need the flow id specifically or is that just to be able to access something else? In the example above the entry is created inside the test so we have total control of what is forwarded.

vangorra replied:
The flow_id is used later to run through the config flow process by calling `hass.config_entries.flow.async_configure` on conftest.py:222 and other places in conftest.

vangorra commented:
What else can be done to get more eyes on this PR to get closer to merging?

riemers commented:
We could ask @balloob nicely (or if he can point someone to this)

frenck commented on CODEOWNERS:
```diff
@@ -258,6 +258,7 @@ homeassistant/components/weather/* @fabaff
 homeassistant/components/weblink/* @home-assistant/core
 homeassistant/components/websocket_api/* @home-assistant/core
 homeassistant/components/wemo/* @sqldiablo
+homeassistant/components/withings/* @vangorra
```
This excludes all tests from the code coverage.
Please only exclude files that don't have tests.

balloob replied:
This is the CODEOWNERS file 👍

frenck replied:
🤦‍♂ I had an awesome week. Sorry about that.

balloob commented:
@riemers please do not comment on PRs tagging people.

balloob commented on homeassistant/components/withings/__init__.py:
```diff
@@ -0,0 +1,114 @@
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Unload Withings config entry."""
+    hass.async_create_task(
```
You can await this. Only during set up do you need to use create_task because of a circular dependency.

balloob commented on homeassistant/components/withings/config_flow.py:
```diff
@@ -0,0 +1,242 @@
+        client_id,
+        client_secret,
+        base_url,
+        profile
```
Why do this instead of having people pick from a list of profiles during config flow ?

balloob commented on homeassistant/components/withings/config_flow.py:
```diff
@@ -0,0 +1,242 @@
+        """Get a profile config entry."""
+        entries = self.hass.config_entries.async_entries(const.DOMAIN)
+        for entry in entries:
+            if entry.title == profile:
```
In the future users might be able to change the title. Compare to data stored in `entry.data` instead.

balloob commented on homeassistant/components/withings/config_flow.py:
```diff
@@ -0,0 +1,242 @@
+
+        url = auth_client.get_authorize_url()
+
+        return self.async_show_form(
```
We should use `self.async_external_step`. For an example, see https://github.com/home-assistant/home-assistant/pull/23782

balloob commented on homeassistant/components/withings/config_flow.py:
```diff
@@ -0,0 +1,242 @@
+
+    async def async_step_code(self, user_input=None):
+        """Received code for authentication."""
+        # if self.hass.config_entries.async_entries(const.DOMAIN):
```
Commented code?

balloob commented on homeassistant/components/withings/config_flow.py:
```diff
@@ -0,0 +1,242 @@
+
+        _LOGGER.debug("Should close all flows below %s",
+                      self.hass.config_entries.flow.async_progress())
+        # Remove notification if no other discovery config entries in progress
```
This comment doesn't apply to anything?

balloob commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,639 @@
+
+    _LOGGER.debug("Confirming we're authenticated")
+    try:
+        data_manager.check_authenticated()
```
Not sure if you're going to add more platforms, but in that case it would make sense to move this to the `async_setup_entry` before forwarding the entry.

vangorra replied:
No intention to add additional platforms.

balloob replied:
We should still check auth in the initial async_setup_entry.

balloob commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,639 @@
+
+    for attribute in WITHINGS_ATTRIBUTES:
+        if attribute.measurement not in measures:
+            _LOGGER.debug("Skipping measurement %s as it is not in the list of measurements to use", attribute.measurement)  # pylint: disable=line-too-long  # noqa: E501
```
Instead of disabling the line too long, just make the line not so long. Python will automatically concatenate multiple strings if there is just whitespace in between.

```python
_LOGGER.debug(
    "Skipping measurement %s "
    "as it is not in the list of measurements", attribute.measurement
)
```

balloob commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,639 @@
+            continue
+
+        _LOGGER.debug(
+            "Creating entity for measurement: %s, measure_type: %s, friendly_name: %s, unit_of_measurement: %s",  # pylint: disable=line-too-long  # noqa: E501
```
Same

balloob commented on homeassistant/components/withings/sensor.py:
```diff
@@ -0,0 +1,639 @@
+        elif unit_of_measurement is const.UOM_LENGTH_M:
+            state = round(value, 2)
+
+        elif unit_of_measurement is const.UOM_LENGTH_CM:
```
We should remove all the custom unit conversions that are done

balloob replied:
An integration should represent the data as-is from the data source. 

If conversion to other units is something that needs to be available for an entity, we should change that in a way that it's available for all entities of a specific device class, not handle it at a per integration level.

pvizeli commented:
Stale. Please reopen after all comments are addressed

isabellaalstrom commented:
@vangorra Are you still working on this one or are you open to someone else trying to fix?

vangorra commented:
Yep. Became distracted with other life events. Will resume soon.

On Fri, Jul 12, 2019, 11:55 PM Isabella Gross Alström <
notifications@github.com> wrote:

> @vangorra <https://github.com/vangorra> Are you still working on this one
> or are you open to someone else trying to fix?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/21625?email_source=notifications&email_token=AANL2MWAQUKCERODE3WQYITP7F375A5CNFSM4G3MMKP2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ3LH5Y#issuecomment-511095799>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANL2MTIQTPEK6ZKKV4N3XDP7F375ANCNFSM4G3MMKPQ>
> .
>

jamiepryer commented:
great news! thank you!

vangorra commented:
Reopened this PR as #25154 with all the requested changes.

southwestflavorz commented:
@vangorra, The in_bed function does not seem to work regardless of if webhook is enabled or not.  Are you experiencing this as well?

MartinHjelmare commented:
Please open an issue if you suspect a bug. If you need help please use our help channels:
https://home-assistant.io/help/#communication-channels

Closed PRs should not be used for support or bug reports. Thanks!
